README.md
percona-chef Cookbook
=====================
This cookbook is to install and manage Percona MySQL 5.6 with xtrabackup. A good example on how to manage the configurations dynamically.

Requirements
------------
Works with Percona 5.6 as I made an old cookbook public. Will update with the latest version as I have more time

Contributing
------------
TODO: Make it work with the latest percona version

e.g.
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: TODO: List authors


CHANGELOG.md
mysql CHANGELOG
===================

# Change log

## master (released)

### New features


test/integration/default/serverspec/mysql_spec.rb
require 'serverspec'

set :backend, :exec

packages = [
  'perl-DBD-MySQL',
  'perl-DBI',
  'Percona-Server-client-56',
  'Percona-Server-devel-56',
  'Percona-Server-server-56',
  'Percona-Server-shared-56',
  'percona-xtrabackup',
  'percona-toolkit']

packages.each do|p|
  describe package(p) do
    it { should be_installed }
  end
end

describe service('mysql') do
  it { should be_enabled   }
  it { should be_running   }
end

describe port(3306) do
  it { should be_listening }
end

describe file('/var/log/mysql-slow.log') do
  it { should be_file }
end

describe file('/etc/my.cnf') do
  it { should be_file }
end

describe file('/root/creds/.mysqlpassword') do
  it { should be_file }
end


recipes/client.rb
#
# Cookbook Name:: percona-chef
# Recipe:: client
#
# Author: Jagatveer Singh
# All rights reserved - Do Not Redistribute
#

%w(mysql-libs
   percona-xtrabackup-20
   percona-xtrabackup-21
   mysql55-5.5.42-1.4.amzn1.x86_64
   mysqlclient16-5.1.61-1.ius.el6.x86_64
).each do |pkg|
  bash "removing #{pkg}" do
    code <<-EOF
          rpm --nodeps -e #{pkg} 2> /dev/null
          exit 0
      EOF
    only_if "rpm -qa | grep #{pkg}"
  end
end

bash 'installing percona' do
  code <<-EOF
        rpm -ivh http://www.percona.com/downloads/percona-release/redhat/0.1-3/percona-release-0.1-3.noarch.rpm
        sed -i 's/$releasever/6/g' /etc/yum.repos.d/percona-release.repo
    EOF
  not_if 'rpm -qa | grep percona-release-0.1-3'
end

# Package perl-DBD-MySQL-4.023-5.17.amzn1.x86_64 is obsoleted by perl-DBD-MySQL55-4.023-5.23.amzn1.x86_64 which is already installed
if node['kernel']['release'][0].to_i >= 4
  dbd_pkg = 'perl-DBD-MySQL55'
else
  dbd_pkg = 'perl-DBD-MySQL'
end

package dbd_pkg do
  action :install
end

%w(perl-DBI
   Percona-Server-client-56
   Percona-Server-devel-56
   Percona-Server-shared-56
   percona-xtrabackup
   percona-toolkit).each do |pkg|
  package pkg do
    action :install
  end
end


libraries/mysql.rb
#
# Cookbook Name:: percona-chef
# Library:: mysql
#
# Author: Jagatveer Singh
# All rights reserved - Do Not Redistribute
#

class Chef::Recipe::MySQL
  def self.get_cnf_by_ram(name, type, ram_gb)
    suffix = type + '.erb'
    if ram_gb > 36
      return name + '-extralarge.' + suffix
    elsif ram_gb > 20
      return name + '-large.' + suffix
    elsif ram_gb > 8
      return name + '-medium.' + suffix
    else
      return name + '-small.' + suffix
    end
  end

  def self.get_innodb_log_size(cnf)
    value = case cnf
            when 'my-extralarge.cnf.erb' then '768M'
            when 'my-large.cnf.erb' then '768M'
            when 'my-medium.cnf.erb' then '512M'
            when 'my-small.cnf.erb' then '256M'
            else ''
            end
    value
  end
  def self.get_innodb_buffer_pool_size(cnf)
    value = case cnf
            when 'my-extralarge.cnf.erb' then '20G'
            when 'my-large.cnf.erb' then '10G'
            when 'my-medium.cnf.erb' then '4G'
            when 'my-small.cnf.erb' then '256M'
            else ''
            end
    value
  end
  def self.get_key_buffer_size(cnf)
    value = case cnf
            when 'my-extralarge.cnf.erb' then '1300M'
            when 'my-large.cnf.erb' then '800M'
            when 'my-medium.cnf.erb' then '500M'
            when 'my-small.cnf.erb' then '300M'
            else ''
            end
    value
  end
end


recipes/xtrabackup.rb
#
# Cookbook Name:: percona-chef
# Recipe:: xtrabackup
#
# Author: Jagatveer Singh#
# All rights reserved - Do Not Redistribute
#

directory node['mysql']['xtrabackup_script_path'] do
  action :create
  recursive true
end

directory node['mysql']['xtrabackup_path'] do
  action :create
  recursive true
end

directory node['mysql']['xtrabackup_base'] do
  action :create
  recursive true
end

directory node['mysql']['xtrabackup_archive'] do
  action :create
  recursive true
end

directory node['mysql']['xtrabackup_mysql_bak'] do
  action :create
  recursive true
end

template node['mysql']['xtrabackup_script_path'] + 'backup.sh' do
  source 'backup.sh.erb'
  mode '755'
  variables('xtrabackup_path' => node['mysql']['xtrabackup_path'],
            'xtrabackup_base' => node['mysql']['xtrabackup_base'],
            'archive' => node['mysql']['xtrabackup_archive'])
end

template node['mysql']['xtrabackup_script_path'] + 'restore.sh' do
  source 'restore.sh.erb'
  mode '755'
  variables('xtrabackup_path' => node['mysql']['xtrabackup_path'],
            'xtrabackup_mysql_bak' => node['mysql']['xtrabackup_mysql_bak'])
end


recipes/default.rb
#
# Cookbook Name:: percona-chef
# Recipe:: default
#
# Author: Jagatveer Singh
# All rights reserved - Do Not Redistribute
#
# https://www.percona.com/forums/questions-discussions/mysql-and-percona-server/percona-server-5-6/22586-yum-repository-seems-to-be-broken

include_recipe 'mysql::client'

package 'Percona-Server-server-56' do
  action :install
end

# Calclutate Amount of RAM present from OHAI plugin
ram_gb  = node['memory']['total'][0..-3].to_f / 1_024_000
my_cnf_file = MySQL.get_cnf_by_ram('my', 'cnf', ram_gb)

bash 'Create mysql slow query log file' do
  not_if { File.exist?('/var/log/mysql-slow.log') }
  code <<-EOH
        set -e
        touch /var/log/mysql-slow.log
        chown mysql.mysql /var/log/mysql-slow.log
        chmod 755 /var/log/mysql-slow.log
    EOH
end

if (node['mysql']['innodb_log_file_size'] == false)
  innodb_log_file_size = MySQL.get_innodb_log_size(my_cnf_file)
else
  innodb_log_file_size = node['mysql']['innodb_log_file_size']
end

if (node['mysql']['innodb_buffer_pool_size'] == false)
  innodb_buffer_pool_size = MySQL.get_innodb_buffer_pool_size(my_cnf_file)
else
  innodb_buffer_pool_size = node['mysql']['innodb_buffer_pool_size']
end

if (node['mysql']['key_buffer_size'] == false)
  key_buffer_size = MySQL.get_key_buffer_size(my_cnf_file)
else
  key_buffer_size = node['mysql']['key_buffer_size']
end

template '/etc/my.cnf' do
  source my_cnf_file
  variables(innodb_log_file_size: innodb_log_file_size,
            innodb_buffer_pool_size: innodb_buffer_pool_size,
            key_buffer_size: key_buffer_size
           )
end

service 'mysql' do
  action [:enable, :start]
end

# Assigning a Secure Password to MySQL
bash 'Generate MySql Password' do
  not_if { File.exist?('/root/creds/.mysqlpassword') }
  code <<-EOH
        set -e
        mysql_password=$(date +%s | sha256sum | base64 | head -c 15 ; echo)
        mysql -u root -e "UPDATE mysql.user SET Password=PASSWORD('$mysql_password') WHERE User='root';FLUSH PRIVILEGES;"
        echo $mysql_password > '/root/creds/.mysqlpassword'
        EOH
end

bash 'clearing the unrequired stuff from percona' do
  code <<-EOH
        mysql -u root -p$(cat /root/creds/.mysqlpassword) -e 'DELETE FROM mysql.user WHERE User="";'
        mysql -u root -p$(cat /root/creds/.mysqlpassword) -e 'DELETE FROM mysql.user WHERE User="root" AND Host NOT IN ("localhost", "127.0.0.1", "::1");'
        mysql -u root -p$(cat /root/creds/.mysqlpassword) -e 'DROP DATABASE test;'
        mysql -u root -p$(cat /root/creds/.mysqlpassword) -e 'FLUSH PRIVILEGES;'
    EOH
end
